Docs for v0.5.0 (#24)
stackrlabs/docs: docs/pages/build/framework/state-machine/block-hooks.mdx

Reviewer: In general feedback
- Please don't give whitespace before `,`
- Don't give whitespace before `:`
- Don't give whitespace before `.`

Author: Will keep in mind

Reviewer 2: the page does not mention how to setup hooks in state machine

Author: Actually here, it was just meant to create hooks. That part is covered in `docs/pages/build/framework/state-machine/introduction.mdx` , it is quite similar to `transitions`

Reviewer: Can we link that to this page too?

Author: Added

> # Block Hooks [Triggered on every block]

Reviewer: ```suggestion
# Block Hooks
```

> Blocks hooks are functions that are automatically executed before and after the execution of a block. Similar to [State transition functions](/build/framework/state-machine/state-transitions) , they also have an handler . They operate at the block level, affecting the entire block's execution.

Reviewer: ```suggestion
Blocks hooks are functions (Similar to [State transition functions](/build/framework/state-machine/state-transitions)) that are automatically executed before or after the execution of all the actions on the block as per configuration. Hooks are executed once per block and a representation of the same is stored in Block Headers for verification later.
```

## Hooks in Stackr

> Stackr SDK also provides a way to create hooks for the blocks.

Reviewer: ```suggestion
Our SDK provides a way to create execution hooks for the blocks.
```

## Hooks in Detail

The most important part of the Hooks is the `handler` function. This function only has access to the current state & the blockInfo `height` , `timestamp` & `parentHash`. Like State Transition Functions , they should be pure functions too & should return the new state after the applying the logic.

:::warning
There is no execution status for the hooks. If a hook execution fails, then the block execution fails too.
:::

## Defining the Hooks for MicroRollup

To configure these hooks, `pre` & `post` hooks can be defined as follows on the MircoRollup :

```ts [mru.ts]
const mru = await MicroRollup({
  config: stackrConfig,
  actionSchemas: [UpdateCounterSchema],
  stateMachines: [machine],
  blockHooks: // [!code hl] {
    pre: ["addTwo"], // [!code hl]
    post: ["minusTwo"], // [!code hl]
  },// [!code hl]
});
```

In this case , we call `addTwo` before and `minusTwo` after the block is executed. Mutliple hooks can be added for both of the triggers by just passing the hook name as string in the array.

:::info

> In case of multiple hooks, Hooks names in the array should be passed in the order of their expected execution

Reviewer: ```suggestion
When dealing with multiple hooks, ensure that the names of the hooks in the array are passed in the order corresponding to their intended execution.
```
